### PR TITLE
feat added local module to downsample bam file

### PR DESCRIPTION
**Pull request by Lucpen**
## PR checklist

- [ ] feat added local module to downsample bam file, it will keep a low percentage of regions within a bed file and then subsample the whole bam file to 80million reads
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/genomic-medicine-sweden/tomte/tree/master/.github/CONTRIBUTING.md)- [ ] If necessary, also make a PR on the nf-core/tomte _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: modules/local/rna_downsample.nf [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+    """
+    samtools view -@ $task.cpus -b -U non_select.bam -L ${downsample_bed} ${bam} | samtools view $seed_frac -@ $task.cpus -b -o select.bam
+    samtools index non_select.bam
+    fraction=\$(samtools idxstats non_select.bam | cut -f3 | awk -v ct="$num_reads" 'BEGIN {total=0} {total += \$1} END {print ct/total}')
+    percent=\$(samtools idxstats non_select.bam | cut -f3 | awk -v ct="$num_reads"00 'BEGIN {total=0} {total += \$1} END {print ct/total}')
+    percent=\$(echo \$percent | cut -d. -f1)
+
+    if (( \$(echo "\$percent < 100") ))
+    then
```
Comment (jemten) [inline modules/local/rna_downsample.nf:37]: Is percent necessary. Can't you use the fraction like this  `if (( \$echo "\$fraction < 1 | bc") ))`? By using `bc` you should be able to do float comparison

File: modules/local/rna_downsample.nf
Comment (jemten) [inline modules/local/rna_downsample.nf:1]: I think we should split this process into two processes.
#### Process 1 
If a downsampling bed file has been given then filter out that region and and downsample it using the `seed_fraction`(parameterized). 
Then we merge the two bam files and index. These files are emitted.

#### Process 2
use samtools idxstats to calculate number of reads and compare it to the wished downsampling depth (parametrized). If possible downsample and index.

This way we can downsample even if we don't want to remove a certain region and we can also get the output from the non-downsampled reads if we want to use them for other processes.

**Review by jemten — COMMENTED** [review]
Nice work. Some comments

File: modules/local/rna_downsample.nf [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    tuple val(meta), path("*_downsmapled.bam"), path("*_downsmapled.bam.bai") , emit: bam_bai
+    path  "versions.yml"                                                      , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def seed_frac       = (args.contains('-s')) ? '' : '-s 0.001'
```
[the remark this answers is not in the document]
Comment (Lucpen) [reply modules/local/rna_downsample.nf]: This is now done :)

File: modules/local/rna_downsample.nf [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    def num_reads       = (args.contains('-num_reads')) ? '' : '80000000'
```
[the remark this answers is not in the document]
Comment (Lucpen) [reply modules/local/rna_downsample.nf]: This is now done :)

File: modules/local/rna_downsample.nf
Comment (Lucpen) [reply modules/local/rna_downsample.nf:1]: Done too :)

File: modules/local/rna_downsample.nf [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    tag "$meta.id"
+    label 'process_low'
+
+    conda "bioconda::samtools=1.16.1"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/samtools:1.16.1--h6899075_1' :
+        'quay.io/biocontainers/samtools:1.16.1--h6899075_1' }"
+
+    input:
+    tuple val(meta), path(bam)
```
Comment (jemten) [inline modules/local/rna_downsample.nf:11]:
```suggestion
    tuple val(meta), path(bam), path(bai)
```

Comment (jemten) [reply modules/local/rna_downsample.nf:11]: idxstats is much quicker if you supply the index for the bam as well (which you already have).
http://www.htslib.org/doc/samtools-idxstats.html

**Review by jemten — APPROVED** [review]
Looks great! 
Only one comment on adding the index to the downsample input.

**Pull request merged**